Add post-processing pipeline for exporting benchmark results to OpenSearch

acalhounRH opened the pull request:
## Summary

This PR adds a post-processing pipeline that converts Zathras benchmark results into structured JSON documents and exports them to OpenSearch for analysis and performance tracking.

## What's Included

**12 Benchmark Processors**
- CPU: CoreMark, CoreMark Pro, SPEC CPU 2017, Passmark
- Memory: STREAMS
- I/O: FIO
- Compute: Auto HPL
- Java: SpecJBB
- Python: PyPerf
- Network: Uperf
- Scheduling: Apache Pig
- Multi-workload: Phoronix Test Suite

**Key Features**
- Object-based schema with full metadata extraction (CPU, memory, NUMA, OS config, cloud provider)
- Two-index design: `zathras-results` (summary) + `zathras-timeseries` (individual points)
- Content-based deduplication (SHA256 hashing prevents duplicate uploads)
- OpenSearch and Horreum export with automatic retry logic
- Batch processing with recursive directory discovery

## Usage

python3 -m post_processing.run_postprocessing \
  --input /path/to/results/ \
  --config export_config.yml \
  --opensearch

- ✅ Tested with 34 Azure instances, 78 benchmarks
- ✅ Exported 629 summary docs + 34,698 time series points
- ✅ Deduplication verified (reprocessing updates, doesn't duplicate)
- ✅ Zero flake8 errors (1,340 issues fixed)

## Files Changed

- **35 files added**, **10,568 lines** (35% code, 65% docs/config)
- Main: `run_postprocessing.py`, `schema.py`, 12 processors
- Exporters: OpenSearch, Horreum, TimeSeries
- Utils: Archive handler, metadata extractor, parsers
- Config: Index templates, example YAML

## new feature addition only.

## Dependencies

- `pyyaml>=6.0`
- `opensearch-py>=2.0.0`
- `requests>=2.28.0`
- `python-dateutil>=2.8.0`

---

## AI Assistance

This PR was developed with AI assistance using **Claude Sonnet 4.5** (Anthropic) via Cursor IDE. The implementation was collaboratively developed through iterative design discussions, code generation, testing, and refinement.

acalhounRH committed:
Add post-processing pipeline for exporting benchmark results to OpenSearch and Horreum

- Implement 12 benchmark processors (CoreMark, STREAMS, FIO, SPEC CPU 2017, etc.)
- Object-based schema with full metadata extraction (CPU, memory, NUMA, OS config)
- Two-index design: zathras-results (summary) + zathras-timeseries (individual points)
- Content-based deduplication using SHA256 hashing
- OpenSearch and Horreum exporters with automatic retry logic
- Batch processing with recursive directory discovery
- Zero flake8 errors (1,340 issues fixed)

Tested with 34 Azure instances, 78 benchmarks successfully exported.

AI-assisted implementation using Claude Sonnet 4.5 (Anthropic) via Cursor IDE.

kdvalin commented on post_processing/exporters/opensearch_exporter.py:
```diff
@@ -0,0 +1,586 @@
+                    data=request_data,
+                    headers=headers,
+                    method=method
+                )
```
Isn't there an opensearch module for python?

We could stay with `requests` to minimize dependencies.  @frival @dvalinrh thoughts?

kdvalin commented on post_processing/utils/archive_handler.py:
```diff
@@ -0,0 +1,372 @@
+Archive Handler Utility
+
+Handles extraction of Zathras result archives:
+- results_{test}.zip → results_{test}_.tar → test result files
```
@dvalinrh did we go back to a regular tarball instead of a zip -> tar -> test results?

kdvalin commented on post_processing/utils/metadata_extractor.py:
```diff
@@ -0,0 +1,569 @@
+                if 'model_name' in proc_data:
+                    cpu_info['model'] = proc_data['model_name']
+                if 'cpu_cores' in proc_data:
+                    cpu_info['cores'] = int(proc_data['cpu_cores'])
```
Same here

kdvalin commented on post_processing/utils/metadata_extractor.py:
```diff
@@ -0,0 +1,569 @@
+        elif unit == 'G':
+            return round(value)
+        elif unit == 'T':
+            return round(value * 1024)
```
Are we getting IEC base-2 amounts or SI base-10 amounts?

Mix/matching them would be probably close enough:tm:, but could provide headaches.

kdvalin reviewed (changes requested):
A quick initial review

grdumas committed:
CoreMark: use CSV timestamps and require them

- Parse CoreMark results CSV with comma delimiter and optional comment
  lines (skip_comments in parser_utils.parse_csv_timeseries).
- Use Start_Date/End_Date from the CSV for run and time-series
  timestamps when present.
- Require valid timestamps: raise ProcessorError instead of synthetic
  timestamps when the CSV lacks Start_Date/End_Date or when a row has
  missing/invalid Start_Date.
- Drop legacy colon-delimited, no-timestamps format support.

grdumas committed:
streams: use CSV timestamps and require valid Start_Date/End_Date

- Parse Start_Date and End_Date from comma-delimited results_streams.csv
  and use them for run timeseries instead of datetime.now().
- Require the CSV header to include Start_Date and End_Date as the last
  two columns; raise ProcessorError with a clear message when timestamps
  are missing, empty, or not valid ISO 8601.
- Support direct CSV path via files['results_streams_csv'] and allow
  results_streams.csv in extracted_path (no streams_* subdir required),
  matching the pattern used under tmp/coremark.

grdumas committed:
autohpl: use timestamps from results, require valid timestamps

- Use timestamps from benchmark output: require comma-delimited CSV with
  Start_Date and End_Date (ISO 8601); use them for run start_time/end_time
  and a single timeseries point. No datetime.now()/utcnow().
- Require timestamps and fail clearly: when timestamps are missing or
  malformed, raise ProcessorError with a short message (expected format
  T/V,N,NB,P,Q,Time,Gflops,Start_Date,End_Date). Add ISO 8601 validation
  helper and reject legacy colon-delimited format without timestamps.

grdumas committed:
coremark_pro: use timestamps from results, require valid timestamps

- Use timestamps from benchmark output: require comma-delimited CSV with
  Start_Date and End_Date (ISO 8601). Use run-level start/end from the
  results and per-workload timestamps from each row's Start_Date for
  timeseries. No datetime.now()/utcnow().
- Require timestamps and fail clearly: when timestamps are missing from
  the header or malformed in any row, raise ProcessorError with a short,
  descriptive message (expected format
  Test,Multi_iterations,Single_iterations,Scaling,Start_Date,End_Date).
  Add ISO 8601 validation helper; reject legacy colon-delimited format
  without timestamps.
- Match tmp/coremark path layout: allow direct results path via
  extracted_result['files']['results_csv'] when provided (demos pass
  single file path). When using extracted_path, allow results.csv
  directly in that directory or in a coremark_pro_* subdirectory.
- Demo: add tmp/coremark_pro/ CSVs (valid, no_timestamps, invalid and
  empty timestamp rows) and post_processing/demo_coremark_pro_timestamps.py
  using the same files-based extracted_result pattern as the CoreMark
  demo; run valid (success), no timestamps (ProcessorError), invalid and
  empty timestamp (ProcessorError) cases with short summary and error
  message output.

grdumas committed:
SpecJBB processor: timestamps from results, require valid, files-based path

- Use timestamps from benchmark output: parse CSV with Start_Date/End_Date
  (ISO 8601) per row and use them for timeseries points instead of
  datetime.now().
- Require timestamps and fail clearly: when CSV lacks Start_Date/End_Date
  or any value is missing/malformed, raise ProcessorError with a short
  message (expected format and what went wrong). Reject legacy
  colon-delimited Warehouses:Bops format without timestamps.
- Add _validate_iso8601_timestamp() and _ISO8601_PATTERN for validation.
- Allow direct results path via extracted_result['files']['results_specjbb_csv']
  when provided; when using extracted_path, look for results_specjbb.csv in
  that directory or via _find_csv_file (rglob).

Assisted-by: Cursor

grdumas committed:
Uperf processor: timestamps from results, require valid, files-based path

- Use timestamps from benchmark output: parse CSV with Start_Date/End_Date
  (ISO 8601) per row for single-CSV path; for net_results layout use
  run_metadata.csv run-level Start_Date/End_Date and interpolate
  timestamps for each data point. No datetime.now()/utcnow().
- Require timestamps and fail clearly: when CSV lacks Start_Date/End_Date
  or any value is missing/malformed, raise ProcessorError with a short
  message (expected format and what went wrong).
- Add _validate_iso8601_timestamp() and _ISO8601_PATTERN for validation.
- Allow direct results path via extracted_result['files']['results_uperf_csv']
  when provided. When using extracted_path, look for results_uperf.csv in
  that directory or in a uperf_* subdir; if not found, use net_results/
  and require run_metadata.csv with Start_Date,End_Date.

Assisted-by: Cursor

grdumas committed:
speccpu2017: use timestamps from results, require valid timestamps, files-based path

- Use timestamps from benchmark output: parse Start_Date/End_Date from CSV
  (format "Benchmarks,Base copies,Base Run Time,Base Rate,Start_Date,End_Date")
  instead of datetime.now()/utcnow(). Per-row timestamps are used for
  timeseries points; run-level start/end are min/max across rows.
- Require timestamps and fail clearly: when CSV lacks Start_Date/End_Date
  (legacy format) or any row has missing/malformed timestamps, raise
  ProcessorError with a short, descriptive message and expected format.
- Add ISO 8601 validation helper (_validate_iso8601_timestamp) in the
  processor; reject blank or non-ISO values.
- Match tmp/coremark layout: allow direct results path via
  extracted_result['files'] (results_csv for single file, or
  results_speccpu_intrate_csv / results_speccpu_fprate_csv). When using
  extracted_path, allow result CSVs in that directory directly or in
  result/ subdirectory.
- Populate Run with start_time, end_time, duration_seconds, and
  timeseries (one TimeSeriesPoint per benchmark row with timestamp from
  Start_Date). Legacy CSV format without timestamps is no longer accepted.

Assisted-by: Cursor

grdumas committed:
passmark: use timestamps from results, require valid timestamps, files-based path

- Use timestamps from benchmark output: read BaselineInfo.TimeStamp from each
  results_all_*.yml (14-digit YYYYMMDDHHmmss), convert to ISO 8601 for
  timeseries points. No datetime.now()/utcnow() fallback.
- Require timestamps and fail clearly: when BaselineInfo.TimeStamp is
  missing, empty, or malformed, raise ProcessorError with a short,
  descriptive message and expected format (YYYYMMDDHHmmss).
- Add format-specific validation helper (_validate_passmark_timestamp) that
  validates 14-digit timestamp and returns ISO 8601 string.
- Match tmp/coremark layout: allow direct results path via
  extracted_result['files']['results_yml'] (single path or list) for demos;
  when using extracted_path, allow results_all_*.yml directly in that
  directory without requiring a test-specific subdirectory.
- Raise ProcessorError when no result files are found or no valid
  iterations (e.g. all YMLs missing Results or timestamp). _build_run_object
  requires a timestamp for each timeseries point.

Assisted-by: Cursor

grdumas committed:
pyperf: use timestamps from results, require valid timestamps, files-based path

- Use timestamps from benchmark output: read metadata.date from each run in
  PyPerf JSON (benchmarks[].runs[].metadata.date). Accept PyPerf format
  (YYYY-MM-DD HH:MM:SS or .ffffff) and ISO 8601 with Z; normalize to ISO 8601.
  No datetime.now()/utcnow() fallback.
- Require timestamps and fail clearly: when metadata.date is missing, empty,
  or malformed, raise ProcessorError with a short, descriptive message and
  expected format. Add _parse_pyperf_timestamp() to validate and normalize.
- Set run-level start_time/end_time from min/max of per-run timestamps in
  the results.
- Allow direct results path via extracted_result['files']['results_json'] or
  ['pyperf_json'] when provided; when using extracted_path, look for
  pyperf_out_*.json directly in that directory. Raise ProcessorError when no
  JSON found or JSON lacks benchmarks array; _parse_json raises on read/parse
  failure.

Assisted-by: Cursor

grdumas committed:
fio: use timestamps from results, require valid timestamps, files-based path

- Use timestamps from benchmark output: require top-level 'timestamp' (Unix
  seconds) in fio-results.json. Use it for run start_time/end_time (end = start
  + max job elapsed) and as base for timeseries. No datetime.now()/utcnow()
  fallback.
- Require timestamps and fail clearly: when 'timestamp' is missing, null,
  non-numeric, or out of reasonable range (2000–2100), raise ProcessorError
  with a short, descriptive message and expected format. Add
  _validate_fio_timestamp() and _fio_timestamp_to_iso().
- Set Run.start_time and Run.end_time from validated timestamp and job
  elapsed times (ISO 8601 UTC).
- Allow direct results path via extracted_result['files']['fio_results_json']
  when provided (e.g. demos under tmp/fio). When using extracted_path, allow
  results in that directory directly (fio_ndisks_*) without requiring
  export_fio_data_* subdirectory.
- _parse_timeseries now takes required test_timestamp (int); base time comes
  from results only.

Assisted-by: Cursor

grdumas committed:
post_processing: reduce Horreum to stubs, focus on OpenSearch

- Replace HorreumExporter implementation with stubs that raise
  NotImplementedError; keep same public interface (url, username, password,
  **kwargs) so run_postprocessing and config remain unchanged.
- Preserve original Horreum implementation as commented-out code at the
  bottom of horreum_exporter.py for future use.
- In run_postprocessing: catch NotImplementedError from Horreum export,
  log a warning and continue; mark exporter as "(stub)" in logs; update
  docstrings and --horreum help to state Horreum is not implemented.
- Update README, config/README, processors/README, __init__.py, and
  requirements.txt to describe OpenSearch as primary and Horreum as stub.

Assisted by Cursor.

grdumas committed:
post_processing: add PCP archive URL to schema and OpenSearch template

- schema.py: Add optional metadata.pcp_archive_url for the URL to the
  Performance CoPilot archive accompanying each test run (location TBD).
  Exclude pcp_archive_url from content hash so setting the URL later does
  not change document identity.
- config/opensearch_index_template.json: Map metadata.pcp_archive_url as
  keyword so the field is indexed when present.

Assisted by Cursor.

grdumas committed:
Add unit/integration tests for post-processing processor timestamps

Replace the post_processing/demos/ scripts with pytest-based tests that
assert timestamp behavior (valid, missing, invalid, empty) for CSV- and
YML-based processors. No external fixture data: tests write minimal
content to tmp_path and call parse_runs.

- tests/post_processing/: conftest (result_dir fixture, run_processor_parse
  helper), and one test module per processor: coremark_pro, streams,
  autohpl, specjbb, uperf, coremark, speccpu2017, passmark (34 tests total).
- Add pytest to post_processing/requirements.txt.
- .gitignore: .venv/, venv/, post_processing/demos/; drop per-file demo
  exclusions and duplicate .gitignore lines.
- README: install and run tests only from project venv; document
  PYTHONPATH=. pytest tests/post_processing/ -v.

Assisted by Cursor.

grdumas committed:
Add Python virtual environment to .gitignore

grdumas committed:
Add mermaid chart of the post-processor code.

grdumas committed:
Add Python bytecode to .gitignore

kdvalin commented on post_processing/run_postprocessing.py:
```diff
@@ -0,0 +1,627 @@
+# Import exporters
+from .exporters.opensearch_exporter import OpenSearchExporter
+from .exporters.timeseries_exporter import TimeSeriesExporter
+from .exporters.horreum_exporter import HorreumExporter
```
Not a right-now problem, but with better python packaging, we should have most of these in their own module's `__init__.py` files..

grdumas commented:
We have moved the functionality contained in this PR over to a new project, [chronicler](https://github.com/redhat-performance/chronicler).
